Add settings for creating and duplicating spaces

martinbrose opened the pull request:
References https://github.com/huggingface/huggingface_hub/issues/1457

**Key Features**

- Modifies hf_api.py to add space-related settings to create_repo.
- Modifies hf_api.py to add additional settings to duplicate_space.

martinbrose commented:
I tried to come up with tests for duplicating spaces.
But tbh, I don't think I have the best approach. I'm getting account rate-limited.
So I believe I'm a little bit too resource-hungry with the test case.

I'll probably leave this with you @Wauplin if you don't mind. You probably have the best idea of how to mock this...

Wauplin commented on src/huggingface_hub/hf_api.py:
```diff
@@ -5357,6 +5391,21 @@ def duplicate_space(
+                Secrets allow to set secret keys or tokens to a Space without hardcoding them.
+                For more details, see https://huggingface.co/docs/hub/spaces-overview#managing-secrets.
+            variables (`List[Dict[str, str]]`, *optional*):
+                Variables allow to set environment variables to a Space without hardcoding them.
```
```suggestion
                A list of public environment variables to set in your Space. Each item is in the form `{"key": ..., "value": ..., "description": ...}` where description is optional.
```

Wauplin commented on src/huggingface_hub/hf_api.py:
```diff
@@ -5357,6 +5391,21 @@ def duplicate_space(
+                the sleep time (value is fixed to 48 hours of inactivity).
+                See https://huggingface.co/docs/hub/spaces-gpus#sleep-time for more details.
+            secrets (`List[Dict[str, str]]`, *optional*):
+                Secrets allow to set secret keys or tokens to a Space without hardcoding them.
```
```suggestion
                A list of secret keys to set in your Space. Each item is in the form `{"key": ..., "value": ..., "description": ...}` where description is optional.
```

Wauplin commented on src/huggingface_hub/hf_api.py:
```diff
@@ -5357,6 +5391,21 @@ def duplicate_space(
+                Choice of Hardware. Example: `"t4-medium"`. See [`SpaceHardware`] for a complete list.
+            storage (`SpaceStorage` or `str`, *optional*):
+                Choice of persistent storage tier. Example: `"small"`. See [`SpaceStorage`] for a complete list.
+            sleepTimeSeconds (`int`, *optional*):
```
```suggestion
            sleep_time (`int`, *optional*):
```

Same here, `sleep_time` would be best :)

martinbrose replied:
Changed this to `sleep_time` throughout. Also updated the test case.

Wauplin reviewed (commented):
That's a very good quality PR, thanks for the work @martinbrose ! I have added comments about renaming/slightly rephrasing but that's it. About the `duplicate_space` tests, I think you can use the exact same approach as the mocked `create_repo` tests. With mock tests you cannot be rate-limited since http requests are not made to the Hub. I'm fine with adding them myself if you prefer. 

Also to complete this PR, would you mind updating the [Manage your Space](https://huggingface.co/docs/huggingface_hub/guides/manage-spaces#a-simple-example-configure-secrets-and-hardware) guide in the docs? You can update them in the [`docs/` folder](https://github.com/huggingface/huggingface_hub/blob/main/docs/source/en/guides/manage-spaces.md). There is a section in a guide _"Bonus: request hardware when creating the Space!"_. I would update it to show how to set hardware + sleep time + secrets + variables at the same time at repo creation. There is also a guide for [duplicating Spaces](https://huggingface.co/docs/huggingface_hub/v0.16.3/en/guides/repository#clone-a-repository-only-for-spaces) that would benefit from a mention of the new arguments you've added. I would keep the existing minimal example and a another complete example just below and explain in a few words why it's useful. Let me know what you think.

Thanks in advance!

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/huggingface_hub/pr_1625). All of your documentation changes will be reflected on that endpoint.

martinbrose commented on src/huggingface_hub/hf_api.py:
```diff
@@ -2365,7 +2365,11 @@ def create_repo(
-        space_hardware: Optional[str] = None,
+        space_hardware: Optional[SpaceHardware] = None,
+        space_storage: Optional[SpaceStorage] = None,
+        space_sleepTimeSeconds: Optional[int] = None,
```
Apologies, a slight oversight on my side. Corrected.

martinbrose commented:
Thanks for the review and the kind recognition.

I addressed all the code review points and added the documentation.
I'll work on the duplication test cases in the next few days.

Wauplin reviewed (commented):
Nice improvement in the docs! Waiting for the final tests for duplicate_space before merging :)

martinbrose commented:
Hi @Wauplin, I think I was overthinking the test case... we are only testing if the API call is correct.

This is now addressed in the additional test case.

Wauplin reviewed (approved):
Great job @martinbrose! Thanks for the last test, it's perfect like this. The PR is fine to merge now :)
